src/build.rs

```diff
@@ -263,3 +265,323 @@ fn build_nasm(root: &Path, vendor_dir: &Path, out_dir: &Path, target_arch: &str,
+    let cxx_wrapper = generate_cxx_wrappers(&decls);
+
+    let rust_output = &format!("{}", rust_wrapper.join(""));
+    let rust_ffi_out = out_dir.join("ffi.rs");
+    std::fs::write(rust_ffi_out, rust_output).unwrap();
+
+    let cxx_output = &format!("{}{}", cxx_header(), cxx_wrapper.join(""));
+    let cxx_out = out_dir.join("mozjpeg_cxx_wrapper.cpp");
+    std::fs::write(cxx_out.clone(), cxx_output).unwrap();
+
+    let mozjpeg_include = src_dir.join("vendor");
+    let config_dir = out_dir.join("include");
+    cc::Build::new()
+        .file(cxx_out)
+        .flag("-Wno-return-type-c-linkage")
+        .include(mozjpeg_include)
+        .include(config_dir)
+        .cpp(true)
+        .compile("mozjpeg_cxx_wrapper");
+}
+
+#[derive(Debug)]
+struct Decl {
+    name: String,
+    is_pub: bool,
+    doc: Vec<String>,
+    attrs: Vec<String>,
+    args: Vec<(String, String)>,
+    ret_ty: Option<String>,
+    lifetime: Option<String>
+}
+
+fn parse_decls(ffi_src: &str) -> Vec<Decl> {
+    // Remove extern "C" { ... } block
+    let input = {
+        let (start_idx, end_idx) = (ffi_src.find('{').unwrap(), ffi_src.rfind('}').unwrap());
+        &ffi_src[start_idx+1..end_idx]
+    };
+    // Remove all whitespace and split `;` to get an Iterator over the decls:
+    let input: Vec<String> = input.split(';').map(ToString::to_string).collect();
+
+    // Parse each declaration and generate wrappers
+    let mut decls = Vec::new();
+    for decl in &input {
+        if decl.trim().is_empty() { continue; }
+        let (doc, attrs) = {
+            let mut doc = Vec::new();
+            let mut attrs = Vec::new();
+            for line in decl.lines() {
+                let line = line.trim();
+                if line.starts_with("///") {
+                    doc.push(line.to_string());
+                } else if line.starts_with("#") {
+                    attrs.push(line.to_string());
+                } else {
+                    break;
+                }
+            }
+            (doc, attrs)
+        };
+        let decl = decl.replace(|c| c == '\n' || c == '\r', "");
+        let is_pub = decl.find("pub fn").is_some();
+        let start_idx = decl.find("fn ").unwrap();
+        let decl = &decl[start_idx..];
+        let name = {
+                let end_idx0 = decl.find('<');
+                let end_idx1 = decl.find('(');
+                let end_idx = match (end_idx0, end_idx1) {
+                    (Some(v), None) => v,
+                    (None, Some(v)) => v,
+                    (Some(i), Some(j)) => i.min(j),
+                    (None, None) => panic!(),
+                };
+                decl["fn ".len()..end_idx].to_string()
+        };
+        match name.as_str() {
+            | "jpeg_float_add_quant_table"
+                | "jpeg_calc_jpeg_dimensions"
+                | "jsimd_fdct_ifast"
+                | "jsimd_can_rgb_ycc"
+                | "jsimd_can_fdct_ifas"
+                => continue,
```

Comment:
These are functions that I just couldn't find in the headers anywhere (Maybe I am just missing a header or using the wrong mozjpeg version?). Usually when writing these kind of automated tools, its rare that they get you 100%, but can be done. It's often just much less work to accept that they get you 99.9% there, and just manually add what doesn't fit the mold of whatever tool you are suing for generating this stuff.